### ISSUE TITLE
bind fist-bump to git commit hooks

### ISSUE THREAD

**Issue by this-oliver** [this-oliver/fist-bump#1]
At the moment, `fist-bump` only runs after the commit has been created. This is because the package depends on the contents of the commit message to bump the version of the package.

```bash
# current version -> 1.0.0

# create a commit message
git commit -m 'feature: adds git hook integration'

# run fist-bump
npx fistbump

# new version -> 1.1.0
# new updated commit message -> '[1.1.0] feature: adds git hook integration'
```

A better user experience would be to automatically run `fist-bump` within the [git hooks](https://git-scm.com/book/en/v2/Customizing-Git-Git-Hooks) as part of the `pre-commit`, `prepare-commit-msg`, `commit-msg` or `post-commit` hooks. If done successfully, users would only have to commit a message with the appropriate format and `fist-bump` would read the new message, bump the version and amend the changes to the commit in question.

```bash
# current version -> 1.0.0

# create a commit message
git commit -m 'feature: adds git hook integration'

# git hook runs fist-bump behind the scenes
# new version -> 1.1.0
# new updated commit message -> '[1.1.0] feature: adds git hook integration'
```

**Comment by platisd** [comment]
What have you tried so far? Any branches or PRs to take a look at?

**Comment by this-oliver** [comment]
> what have I tried so far?

This tool was originally a [bash script](https://github.com/this-oliver/fist-bump/blob/main/scripts/bump.sh) so I mostly played around with the [simple-git-hooks](https://github.com/toplenboren/simple-git-hooks) tool which binds a command to a git hook.

The problem with the hooks was that I wasn't able to read the commit message in any of the hooks that allowed me to make changes to the commit before it was created (`pre-commit` & `prepare-commit-msg`). The next best solution was the `commit-msg` hook but this one would stage the changes to git but would not amend the previous commit which is the commit that the version bump was meant for.

> any branches or PRs to look at?

🔗 -> [made a quick one for this issue](https://github.com/this-oliver/fist-bump/pull/2)

**Comment by platisd** [comment]
> After the entire commit process is completed, the `post-commit` hook runs. It doesn’t take any parameters, but you can easily get the last commit by running git log -1 HEAD. Generally, this script is used for notification or something similar 

How come `post-commit` doesn't work? Looking at the bash script it looks like it should. :thinking:

**Comment by this-oliver** [comment]
TLDR; running the script in the `post-commit` ignites an endless loop.

Because the script make changes to the version of the `package.json` and then amends the changes to the latest commit, if it has the necessary format.

One possible solution would be to run the [`--no-verify`](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---no-verify) flag which skips the commit hooks but this flag only applies to the `pre-commit` and `commit-msg`. 

All the other hooks are run regardless of the flag, including the `post-commit` hook which means that the script runs -> makes changes -> commits changes as an amendment -> triggers the post-commit hook -> commits new version changes as an amendment -> repeats cycle.

**Comment by platisd** [comment]
> One possible solution would be to run the [--no-verify](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---no-verify) flag which skips the commit hooks but this flag only applies to the pre-commit and commit-msg.

Oh shi, I didn't know :exploding_head: 

What if you temporarily `touch` an empty file (could even be in `/tmp` if we assume a Linux system), like a "flag" after you update the commit message?
If your script runs and finds that file it means that it should remove that file and immediately stop, thus breaking the "loop".

**Comment by this-oliver** [comment]
@platisd you are a x10 engineer XD. This specific commit has been bugging me for a while but I didn't know how to approach it even with your suggestion but I just learnt about env variables (in a different project) and if I combine their behaviour with your 'flag' suggestion, things might actually work out ;)
